### PR TITLE
Add cassandra storage

### PR DESCRIPTION
**Pull request by adejanovski**
Adds Cassandra as a storage option for reaper internal data.
I've also added checks so that we cannot run two concurrent repair sessions for the same repair unit, and so that we cannot create two different repair schedules that have the same table in it.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1,156 +1,167 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (michaelsembwever) [inline pom.xml:1]: non-blocking:
 the diff of this file is hard to read with the reformatting (i presume from your IDE) changing more of the file than is explicit to this PR. it's help a lot if such reformatting changes are kept isolated in separate commits, or if you can disable reformatting in your IDE.

File: src/main/db/reaper_cassandra_db.cql [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  id              bigint PRIMARY KEY,
+  cluster_name    text,
+  keyspace_name   text,
+  column_families set<text>,
+  incremental_repair boolean
+);
+
+CREATE TABLE IF NOT EXISTS repair_run (
+  id                 bigint PRIMARY KEY,
+  cluster_name       TEXT,
```
Comment (michaelsembwever) [inline src/main/db/reaper_cassandra_db.cql]: nit: sometimes text is uppercase, sometimes its lowercase…? same with int.

File: src/main/java/com/spotify/reaper/resources/CommonTools.java [hunk 3]
```diff
@@ -252,6 +253,25 @@ public static RepairSchedule storeNewRepairSchedule(
                                    repairParallelism, intensity,
                                    DateTime.now());
     scheduleBuilder.owner(owner);
+    
+    Collection<RepairSchedule> repairSchedules = context.storage.getRepairSchedulesForClusterAndKeyspace(repairUnit.getClusterName(), repairUnit.getKeyspaceName());
+    for(RepairSchedule sched:repairSchedules){
+    	Optional<RepairUnit> repairUnitForSched = context.storage.getRepairUnit(sched.getRepairUnitId());
```
Comment (michaelsembwever) [inline src/main/java/com/spotify/reaper/resources/CommonTools.java]: blocking: replace tabs with whitespace. (happens in many of the edited files in this PR)

File: src/main/java/com/spotify/reaper/resources/CommonTools.java [hunk 4]
```diff
@@ -252,6 +253,25 @@ public static RepairSchedule storeNewRepairSchedule(
...
+    	if(repairUnitForSched.isPresent() && repairUnitForSched.get().getClusterName().equals(repairUnit.getClusterName()) && repairUnitForSched.get().getKeyspaceName().equals(repairUnit.getKeyspaceName())){
+    		if((repairUnitForSched.get().getColumnFamilies().size()==0 && repairUnit.getColumnFamilies().size()==0)
+    				|| repairUnitForSched.get().getColumnFamilies().size()==0 && repairUnit.getColumnFamilies().size()!=0
+    				|| repairUnitForSched.get().getColumnFamilies().size()!=0 && repairUnit.getColumnFamilies().size()==0
+    				|| Sets.intersection(repairUnit.getColumnFamilies(),repairUnitForSched.get().getColumnFamilies()).size() >0){
```
Comment (michaelsembwever) [inline src/main/java/com/spotify/reaper/resources/CommonTools.java]: non-blocking:
could be made easier to read: if these two if statement conditionals were made into methods with intuitive descriptive names.

eg (names could be better…)

```
if (isEqual(repairUnitForSched, repairUnit) {
    if (isRepairInProgress(repairUnitForSched, repairUnit)) {
            String errMsg = String.format(
                        "A repair schedule already exists for cluster \"%s\",keyspace \"%s\", and column families: %s",
                        cluster.getName(), 
                        repairUnit.getKeyspaceName(),
                        Sets.intersection(repairUnit.getColumnFamilies(),repairUnitForSched.get().getColumnFamilies()));
                    LOG.error(errMsg);
                    throw new ReaperException(errMsg);
        }
}

private static boolean isValid(Optional<RepairUnit> repairUnitForSched, RepairUnit repairUnit) {
    return repairUnitForSched.isPresent() 
            && repairUnitForSched.get().getClusterName().equals(repairUnit.getClusterName()) 
            && repairUnitForSched.get().getKeyspaceName().equals(repairUnit.getKeyspaceName());
}

private static boolean isRepairInProgress(Optional<RepairUnit> repairUnitForSched, RepairUnit repairUnit) {
    return (repairUnitForSched.get().getColumnFamilies().size()==0 && repairUnit.getColumnFamilies().size()==0)
            || repairUnitForSched.get().getColumnFamilies().size()==0 && repairUnit.getColumnFamilies().size()!=0
            || repairUnitForSched.get().getColumnFamilies().size()!=0 && repairUnit.getColumnFamilies().size()==0
            || Sets.intersection(repairUnit.getColumnFamilies(),repairUnitForSched.get().getColumnFamilies()).size() >0;
}
```

File: src/main/java/com/spotify/reaper/resources/RepairRunResource.java [hunk 5]
```diff
@@ -275,6 +277,20 @@ public Response modifyRunState(
       return Response.status(Response.Status.NOT_FOUND).entity(errMsg).build();
     }
 
+    // Check that no other repair run exists with a status different than DONE for this repair unit
+    Collection<RepairRun> repairRuns = context.storage.getRepairRunsForUnit(repairRun.get().getRepairUnitId());
+    
+    boolean alreadyRunning = false;
```
Comment (michaelsembwever) [inline src/main/java/com/spotify/reaper/resources/RepairRunResource.java]: where is this used?

File: src/main/java/com/spotify/reaper/storage/CassandraStorage.java [hunk 6]
```diff
@@ -0,0 +1,696 @@
...
+	@Override
+	public Collection<RepairRunStatus> getClusterRunStatuses(String clusterName, int limit) {
+		// TODO Auto-generated method stub
+		return null;
+	}
+
+	@Override
+	public Collection<RepairScheduleStatus> getClusterScheduleStatuses(String clusterName) {
+		// TODO Auto-generated method stub
+		return null;
```
Comment (michaelsembwever) [inline src/main/java/com/spotify/reaper/storage/CassandraStorage.java]: forgotten?

File: src/main/db/reaper_cassandra_db.cql (hunk 2 above)
Comment (adejanovski) [reply src/main/db/reaper_cassandra_db.cql]: Indeed. I fixed formatting to be consistent

File: src/main/java/com/spotify/reaper/resources/CommonTools.java (hunk 3 above)
Comment (adejanovski) [reply src/main/java/com/spotify/reaper/resources/CommonTools.java]: Changed to spaces by default.

File: src/main/java/com/spotify/reaper/resources/CommonTools.java (hunk 4 above)
Comment (adejanovski) [reply src/main/java/com/spotify/reaper/resources/CommonTools.java]: Right, I refactored the condition into a separate method.
Good catch

File: src/main/java/com/spotify/reaper/resources/RepairRunResource.java [hunk 7]
```diff
@@ -62,6 +63,7 @@
   private static final Logger LOG = LoggerFactory.getLogger(RepairRunResource.class);
 
   private final AppContext context;
+  private Boolean alreadyRunning;
```
[the remark this answers is not in the document]
Comment (adejanovski) [reply src/main/java/com/spotify/reaper/resources/RepairRunResource.java]: Removed

File: src/main/java/com/spotify/reaper/resources/RepairRunResource.java (hunk 5 above)
Comment (adejanovski) [reply src/main/java/com/spotify/reaper/resources/RepairRunResource.java]: Removed

File: src/main/java/com/spotify/reaper/service/SchedulingManager.java [hunk 8]
```diff
@@ -30,7 +30,7 @@ public static void start(AppContext context) {
       LOG.info("Starting new SchedulingManager instance");
       schedulingManager = new SchedulingManager(context);
       Timer timer = new Timer("SchedulingManagerTimer");
-      timer.schedule(schedulingManager, 1000, 1000 * 60 * 10); // activate once per ten minutes
+      timer.schedule(schedulingManager, 1000, 1000 * 60); // activate once per ten minutes
```
[the remark this answers is not in the document]
Comment (adejanovski) [reply src/main/java/com/spotify/reaper/service/SchedulingManager.java]: Alas, it is not 😄 
Removed

File: src/main/java/com/spotify/reaper/storage/CassandraStorage.java (hunk 6 above)
Comment (adejanovski) [reply src/main/java/com/spotify/reaper/storage/CassandraStorage.java]: I did forgot those two indeed 🤔 
They are implemented now, thanks !

**Comment by michaelsembwever** [comment]
LGTM.

remember to collapse (`git rebase -i HEAD~9`)  your commits before merging.
 along with rewriting the collapsed commit it's really helpful to link from the commit msg back to this PR.

**Comment by rustyrazorblade** [comment]
Easiest to merge with `--squash`, it does it for you.

**Comment by adejanovski** [comment]
Squashed + rebased

**Comment by michaelsembwever** [comment]
LGTM

**Comment by zznate** [comment]
Oops. Looks like this butted heads with pull 3. @adejanovski can you rebase and double check the `pom.xml` merge?

**Comment by adejanovski** [comment]
Rebased and fixed conflict. 
You can merge now.

**Pull request merged**